controllers/dockermachine_controller.go

```diff
@@ -91,137 +132,138 @@ func (r *DockerMachineReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
+	}
+
+	// Handle non-deleted machines
+	return r.reconcileNormal(machine, dockerMachine, externalMachine, externalLoadBalancer)
+}
+
+func (r *DockerMachineReconciler) reconcileNormal(machine *clusterv1.Machine, dockerMachine *infrav1.DockerMachine, externalMachine *docker.Machine, externalLoadBalancer *docker.LoadBalancer) (ctrl.Result, error) {
 	// If the DockerMachine doesn't have finalizer, add it.
 	if !util.Contains(dockerMachine.Finalizers, infrav1.MachineFinalizer) {
 		dockerMachine.Finalizers = append(dockerMachine.Finalizers, infrav1.MachineFinalizer)
 	}
-	state := getState(machine, dockerMachine)
-	log = log.WithValues("state", state.String())
-	// TODO: consider setting the key pieces in this function, such as ProviderID and deleting finalizers
-	switch state {
-	case Deleted:
-		log.Info("Start reconcileDelete dockerMachine")
-		return r.reconcileDelete(ctx, cluster, machine, dockerMachine)
-	case Provisioned:
-		return ctrl.Result{}, nil
-	case Pending:
-		log.Info("Waiting for machine bootstrap")
+
+	// if the machine is already provisioned, return
+	if dockerMachine.Spec.ProviderID != nil {
 		return ctrl.Result{}, nil
-	case Provisioning:
-		log.Info("Provisioning machine")
-		// Ensuring cluster is ready for joining
-		clusterExists, err := kindcluster.IsKnown(cluster.Name)
-		if err != nil {
-			log.Error(err, "Error finding cluster-name", "cluster", cluster.Name)
-			return ctrl.Result{}, err
-		}
+	}
 
-		// If there's no cluster, requeue the request until there is one
-		if !clusterExists {
-			r.Log.Info("There is no cluster yet, waiting for a cluster before creating machines")
-			return ctrl.Result{RequeueAfter: time.Second * 30}, nil
-		}
-		log.Info("Creating machine")
-		return r.create(ctx, cluster, machine, dockerMachine)
-	default:
-		log.Info("Unknown state", "state", state)
+	// Make sure bootstrap data is available and populated.
+	if machine.Spec.Bootstrap.Data == nil {
+		externalMachine.Info("Waiting for the Bootstrap provider controller to set bootstrap data")
 		return ctrl.Result{}, nil
 	}
-}
 
-// State is the state our machine object is in
-type State int
+	// Determine if the machine is the first control plane joining the cluster
+	isFirstControlPlane := util.IsControlPlaneMachine(machine) //&& m.Cluster.Annotations[clusterv1.ClusterAnnotationControlPlaneReady] != "true"
 
-const (
-	// Provisioned is the state that happens after the machine is part of a cluster.
-	// Technically it should be Running but this controller sees no difference between the two.
-	Provisioned State = iota
-	// Pending is when the machine is waiting for bootstrap data to exist.
-	Pending
-	// Deleted is when the machine has been deleted.
-	Deleted
-	// Provisioning is when bootstrap data exists but it's not finished being provisioned.
-	Provisioning
-)
+	//Create the docker container hosting the load balancer
+	if util.IsControlPlaneMachine(machine) {
```

Comment:
One option you've got here is to move the creation of `externalMachine` to here and have it take another parameter, `role string` which will either being `controlplane` or `worker`. Then the `Create()` function can hide the fact that it's creating a control plane or worker and the reconciler can just call `externalMachine.Create()`

Reply:
+1